src/hotspot/share/runtime/flags/jvmFlag.hpp

```diff
@@ -141,87 +166,97 @@ struct JVMFlag {
+  void assert_type(int type_enum) const {
+    if (type_enum == JVMFlag::TYPE_ccstr) {
+      assert(is_ccstr(), "type check"); // ccstr or ccstrlist
+    } else {
+      assert(_type == type_enum, "type check");
+    }
+  }
 
-  bool is_double() const;
-  double get_double() const                   { return *((double*) _addr); }
-  void set_double(double value) const         { *((double*) _addr) = value; }
+  // Do not use JVMFlag::read() or JVMFlag::write() directly unless you know
+  // what you're doing. Use FLAG_SET_XXX macros or JVMFlagAccess instead.
+  template <typename T, int type_enum> T read() const {
+    assert_type(type_enum);
+    return *static_cast<T*>(_addr);
+  }
 
-  bool is_ccstr() const;
-  bool ccstr_accumulates() const;
-  ccstr get_ccstr() const                     { return *((ccstr*) _addr); }
-  void set_ccstr(ccstr value) const           { *((ccstr*) _addr) = value; }
+  template <typename T, int type_enum> void write(T value) {
+    assert_type(type_enum);
+    *static_cast<T*>(_addr) = value;
+  }
 
-  Flags get_origin() const;
+  Flags get_origin() const        {  return Flags(_flags & VALUE_ORIGIN_MASK);   }
   void set_origin(Flags origin);
 
-  bool is_default() const;
-  bool is_ergonomic() const;
-  bool is_jimage_resource() const;
-  bool is_command_line() const;
-  void set_command_line();
-
-  bool is_product() const;
-  bool is_manageable() const;
-  bool is_diagnostic() const;
-  bool is_experimental() const;
-  bool is_notproduct() const;
-  bool is_develop() const;
+  bool is_default() const         { return (get_origin() == DEFAULT);            }
+  bool is_ergonomic() const       { return (get_origin() == ERGONOMIC);          }
+  bool is_command_line() const    { return (_flags & ORIG_COMMAND_LINE) != 0;    }
+  void set_command_line()         {  _flags = Flags(_flags | ORIG_COMMAND_LINE); }
+  bool is_jimage_resource() const { return (get_origin() == JIMAGE_RESOURCE);    }
+  bool is_product() const         { return (_flags & KIND_PRODUCT) != 0;         }
+  bool is_manageable() const      { return (_flags & KIND_MANAGEABLE) != 0;      }
+  bool is_diagnostic() const      { return (_flags & KIND_DIAGNOSTIC) != 0;      }
+  bool is_experimental() const    { return (_flags & KIND_EXPERIMENTAL) != 0;    }
+  bool is_notproduct() const      { return (_flags & KIND_NOT_PRODUCT) != 0;     }
+  bool is_develop() const         { return (_flags & KIND_DEVELOP) != 0;         }
 
   bool is_constant_in_binary() const;
 
   bool is_unlocker() const;
   bool is_unlocked() const;
-  bool is_writeable() const;
-  bool is_external() const;
+
+  // Only manageable flags can be accessed by writeableFlags.cpp
+  bool is_writeable() const       { return is_manageable();                      }
+  // All flags except "manageable" are assumed to be internal flags.
+  bool is_external() const        { return is_manageable();                      }
```

Comment:
Do we need is_writeable() and/or is_external() if all they do is return is_manageable() ?

Reply:
is_external() is odd and only has one caller.  You could just fix the caller and remove is_external.

Reply:
Since I am doing a "syntactic" clean up, I want to keep the code behavior the same, without doing any "semantic" changes. The caller of `is_external()` may have a (historically) different intention than the caller of `is_manageable()`. I want to leave such clean up in separate RFEs.

Reply:
Makes sense